### PR TITLE
HDDS-6701. [Multi-Tenant] Add proper locking between Ranger background sync service and tenant requests; bug fixes

### PR DESCRIPTION
**Pull request by smengcl**
## What changes were proposed in this pull request?

- Add proper locking between Ranger Background Sync Service and tenant requests
- Address (most) small bugs filed in https://issues.apache.org/jira/issues/?jql=labels%20%3D%20ozone-multitenancy.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6701

## How was this patch tested?

- [x] Existing tests should pass.

### CODE REVIEW

**Review by errose28 — COMMENTED** [review]
Thanks for working on this @smengcl. Just some minor comments.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMMultiTenantManagerImpl.java [hunk 1]
```diff
@@ -183,246 +197,456 @@ public OMMetadataManager getOmMetadataManager() {
...
+        final CachedTenantState cachedTenantState = tenantCache.get(tenantId);
+        Preconditions.checkNotNull(cachedTenantState,
+            "Cache entry for tenant '" + tenantId + "' does not exist");
+
+        LOG.info("Adding to cache: user '{}' accessId '{}' in tenant '{}'",
+            userPrincipal, accessId, tenantId);
+        cachedTenantState.getAccessIdInfoMap().put(accessId, cacheEntry);
+      } catch (Exception e) {
+        // Attempt to clean up tenant cache entry on exception
+        tenantCache.get(tenantId).getAccessIdInfoMap().remove(accessId);
```
[the remark this answers is not in the document]
Comment (smengcl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMMultiTenantManagerImpl.java:575]: Good point. Removed

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/multitenant/OMRangerBGSyncService.java [hunk 2]
```diff
@@ -273,9 +280,18 @@ public BackgroundTaskResult call() {
   }
 
   private void triggerRangerSyncOnce() {
-    int attempt = 0;
+    long lockStamp = 0L;
     try {
-      // TODO: Acquire lock
+      // Acquire authorizer (Ranger) write lock
```
[the remark this answers is not in the document]
Comment (smengcl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/multitenant/OMRangerBGSyncService.java:293]: I agree. Removed read lock

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMMultiTenantManagerImpl.java [hunk 3]
```diff
@@ -183,246 +197,456 @@ public OMMetadataManager getOmMetadataManager() {
...
+            accessId, tenantId);
+        if (!tenantCache.get(tenantId).getAccessIdInfoMap()
+            .containsKey(accessId)) {
+          throw new OMException("accessId '" + accessId + "' doesn't exist "
+              + "in tenant cache!", INTERNAL_ERROR);
+        }
+        tenantCache.get(tenantId).getAccessIdInfoMap().remove(accessId);
+      } catch (NullPointerException e) {
+        // tenantCache is somehow empty. Ignore for now.
+        LOG.warn("NPE when removing accessId from cache", e);
```
[the remark this answers is not in the document]
Comment (smengcl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMMultiTenantManagerImpl.java:597]: Actually, NPE shouldn't be thrown here anymore.
Tenant existence is checked in preExecute. Removed

**Comment by smengcl** [comment]
Thanks @errose28 for the review. I have addressed the comments. Please take another look, thx!

**Comment by smengcl** [comment]
Thanks @prashantpogde @errose28 for the review.

**Pull request merged**